Use page cache for all operation on record store file during store copy

burqen opened the pull request:
Builds on #7896, new code start at `e297132` (  StoreCopyServer now read from paged file is mapping exists). Will be rebased after #7896 is merged.

When storing the record storage file on some storage external to the underlying file system, only the page cache know how to write to and move those files.

Summary of changes:
- StoreCopyServer read file through PageCache if mapping exists
- ToFileStoreWriter will write through PageCache if file is considered to be a record store file
- Add record size to store copy protocol to allow new mapping of record store files
- Use page cache to move record store files from temp directory to store directory after store copy

Commit:
StoreCopyServer now read from paged file is mapping exists

When performing a store copy, StoryCopyServer will now try to get the ByteChannel from the mapped PagedFile if a mapping exist, otherwise will behave as before and get the ByteChannel from the FileSystem.

 This makes it possible to support store copy from a device external to the normal file system.

Commit:
StoreCopyClient now writes through provided PageCache if file is a store file

When performing a store copy, if file is a store file, StoryCopyClient will now map the target file with the PageCache and write to the file through an opened ByteChannel on the paged file.

 If file is not a store file, then behavior is same as before.

 This makes it possible to support store copy to a device external to the normal file system.

Commit:
StoreType make use of Optional functionality

Commit:
ToFileStoreWriter now write everything in data channel

Commit:
Create store copy directory even if writing through page cache

Commit:
NeoStoreFileListing now use StorageEngine to list store files

And also include meta data about the files, such as StoreType and recordSize.

Commit:
Added record size to store copy protocol

Commit:
Use tuningConfiguration when creating page cache in backup

Commit:
Bug fix. Store copy protocol now write and read recordSize properly

Commit:
StoreCopyServer now use page cache to move files if written through page cache

Commit:
Backup is now allowed with customized IO

Commit:
StoreCopyServer now report hasData correctly depending on file size

Add fileSize() to PagedFile interface

Commit:
Cleanup after messy rebase

chrisvest commented:
Rebased. Marking it as ready for review & merge. /cc @burqen @tinwelint

chrisvest commented on community/io/src/main/java/org/neo4j/io/pagecache/PagedFile.java:
```diff
@@ -131,6 +131,11 @@
     /**
+     * Size of file, in bytes.
+     */
+    long fileSize() throws IOException;
```
There should be tests for this in `PageCacheTest`

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreType.java:
```diff
@@ -227,15 +228,28 @@ void close( NeoStores me, Object object )
+     * @param fileName - name of the file to check for corresponding store type
+     * @return {@link Optional} wrapping store type, or empty optional if no store type corresponds to file name
+     */
+    public static Optional<StoreType> isStoreType( String fileName )
```
Maybe this should be called `findStoreType` or `determineStoreType`, or `guessStoreType`.

chrisvest commented on enterprise/com/src/main/java/org/neo4j/com/storecopy/StoreCopyServer.java:
```diff
@@ -154,13 +159,32 @@ public RequestContext flushStoresAndStreamStoreFiles( String triggerName, StoreW
-                        monitor.finishStreamingStoreFile( file );
+                        try ( ReadableByteChannel fileChannel = fileSystem.open( file, "r" ) )
+                        {
+                            long fileSize = file.length();
```
`fileSystem.getFileSize( file )` ?

chrisvest commented on enterprise/com/src/main/java/org/neo4j/com/storecopy/ToFileStoreWriter.java:
```diff
@@ -23,46 +23,77 @@
     private final File basePath;
     private final StoreCopyClient.Monitor monitor;
+    private final PageCache pageCache;
+    private final List<FileMoveAction> fileMoveActions; // Todo need thread safe?
```
No, I don't think this needs to be thread safe. The files are streaming over one at a time, controlled on our side by a single thread.

chrisvest commented on enterprise/com/src/main/java/org/neo4j/com/storecopy/ToFileStoreWriter.java:
```diff
@@ -23,46 +23,77 @@
+                else
+                {
+
+                    return writeDataThroughFileSystem( file, data, temporaryBuffer, hasData );
```
Why don't this add a move action? Then we'd only need one loop in `StoreCopyClient`.

chrisvest reviewed (commented):
I had a few comments on a look through.

chrisvest reviewed (commented):
Had a few comments

chrisvest committed:
Add tests for PagedFile.fileSize

chrisvest committed:
Address a couple of review comments

MishaDemianenko commented on enterprise/com/src/main/java/org/neo4j/com/storecopy/ToFileStoreWriter.java:
```diff
@@ -75,6 +106,64 @@ public long write( String path, ReadableByteChannel data, ByteBuffer temporaryBu
+    {
+        final int pageCacheSize = pageCache.pageSize();
+        return recordSize == NO_RECORD_SIZE ? pageCacheSize
+                                            : pageCacheSize - pageCacheSize % recordSize;
```
i really miss clarifying parentheses here

MishaDemianenko commented on enterprise/com/src/main/java/org/neo4j/com/storecopy/ToNetworkStoreWriter.java:
```diff
@@ -56,9 +56,10 @@ public long write( String path, ReadableByteChannel data, ByteBuffer temporaryBu
         if ( hasData )
         {
+            targetBuffer.writeInt( recordSize );
+            totalWritten += 4;
```
Integer.BYTES please

MishaDemianenko commented on enterprise/com/src/main/java/org/neo4j/com/storecopy/ToNetworkStoreWriter.java:
```diff
@@ -56,9 +56,10 @@ public long write( String path, ReadableByteChannel data, ByteBuffer temporaryBu
         long totalWritten = 2 + chars.length*2 + 1;
         if ( hasData )
         {
+            targetBuffer.writeInt( recordSize );
```
Shouldn't we update protocol version as result of doing this?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/kvstore/AbstractKeyValueStore.java:
```diff
@@ -248,13 +252,9 @@ private boolean transfer( EntryVisitor<WritableBuffer> producer, EntryVisitor<Re
-                existingFiles.add( candidateFile );
-            }
-        }
+        StreamSupport.stream( rotationStrategy.candidateFiles().spliterator(), false )
```
i think all files can be just this:

```
        return StreamSupport.stream( rotationStrategy.candidateFiles().spliterator(), false )
                .filter( File::exists )
                .collect( Collectors.toList() );
```

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreType.java:
```diff
@@ -227,15 +228,28 @@ void close( NeoStores me, Object object )
      */
     public static StoreType typeOf( String storeFileName )
     {
+        final Optional<StoreType> optional = findStoreType( storeFileName );
```
what if we will just update typeOf(filename) to return Optional instead?

chrisvest replied:
Of course.

MishaDemianenko commented on enterprise/com/src/main/java/org/neo4j/com/storecopy/ToFileStoreWriter.java:
```diff
@@ -23,46 +23,77 @@
+
+            String filename = file.getName();
+            final Optional<StoreType> storeType = findStoreType( filename );
+            final Optional<PagedFile> existingMapping = pageCache.getExistingMapping( file );
```
I think it would be useful to have a comment explaining here what files we expect to be mapped, not mapped, not a stores to simplify future investigations

chrisvest replied:
That's a really good question. In `ToFileStoreWriter` we don't actually need it because we always write to a temporary location. The call in `StoreCopyServer` is seeing more use, so I'm adding a comment there.

chrisvest committed:
Various cleanups around store copying

chrisvest committed:
Remove existing-mapping check from ToFielStoreWriter

It would always fail anyway because we are copying to a temporary location.

chrisvest commented:
@MishaDemianenko I have addressed your comments.

Commit:
Magic numbers

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreType.java:
```diff
@@ -221,21 +222,21 @@ void close( NeoStores me, Object object )
         {
-            if ( value.getStoreName().equals( storeFileName ) ||
-                    storeFileName.equals( MetaDataStore.DEFAULT_NAME + value.getStoreName() ) )
+            if ( value.getStoreName().equals( fileName ) ||
```
To match both (w/ and w/o `neostore.` in front) looks like a source of confusion and a potential edge-case problems finding unrelated files. It'd be better to decide exactly what it looks for, document it and have all callers adhere to it.

MishaDemianenko commented:
@chrisvest thx, all fixes looks nice so far!

tinwelint commented on enterprise/com/src/main/java/org/neo4j/com/storecopy/ToNetworkStoreWriter.java:
```diff
@@ -45,20 +45,21 @@ public ToNetworkStoreWriter( ChannelBuffer targetBuffer, Monitors monitors )
+        long totalWritten = Short.BYTES + chars.length* Character.BYTES + Byte.BYTES;
         if ( hasData )
         {
+            targetBuffer.writeInt( recordSize );
```
It looks like you're changing the network protocol. For compatibility, and detection of incompatibility you have to increment protocol version. I think this is a case where the internal protocol version should be incremented, see https://github.com/neo4j/neo4j/blob/3.1/enterprise/com/src/main/java/org/neo4j/com/ProtocolVersion.java#L24

chrisvest replied:
Yes, I had @digitalstain explain to me in vague terms how this is done.

chrisvest committed:
Clarify alignment field in store copy protocol

Rename the 'recordSize' to 'requiredElementAlignment' in the store copy protocol, to abstract ourselves from the fact that we occasionally copy over record stores with that code.
The code is mean to be generic, and is indeed used to copy over other files, such as indexes and logs, and an element alignment requirement is a generic term for what we want.

chrisvest committed:
Increment the HA protocol version

We have to do this because we've changed the wire format of the file transfer message.

chrisvest committed:
Fix license headers

chrisvest committed:
Clean up and lambda-fy HA code

chrisvest committed:
Allow HA to boot up with custom IO subsystem configurations

chrisvest committed:
Remove the old MasterClient210 and Protocol201

We only (pretend?) to support one version back, so we can remove a lot of the old code but keep the protocol versioning infrastructure.

chrisvest committed:
Simplify StoreType.typeOf, so it now only matches exact store file names

chrisvest commented:
@tinwelint I think I've addressed all of your comments.

tinwelint commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/MasterClient214.java:
```diff
@@ -19,32 +19,78 @@
+     * Version 5 since ?
+     * Version 6 since 2014-01-07
+     * Version 7 since 2014-03-18
+     * Version 8 since 2014-08-27
```
A legend like this is always good to have, this could be moved to MasterClient310 (which is now the latest) and have `version 9` included as well

tinwelint commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToSlave.java:
```diff
@@ -561,7 +561,7 @@ private MasterClient newMasterClient( URI masterUri, URI me, StoreId storeId, Li
-        if ( masterClient.getProtocolVersion().compareTo( MasterClient210.PROTOCOL_VERSION ) < 0 )
+        if ( masterClient.getProtocolVersion().compareTo( MasterClient214.PROTOCOL_VERSION ) < 0 )
         {
             idGeneratorFactory.enableCompatibilityMode();
```
Since we now have removed `MasterClient210`, then `HaIdGeneratorFactory#enableCompatibilityMode()` should also go since it was a work-around for cross-2.0->2.1 cluster migrations. `MasterClient214` has a misleading name, it's when it was created, but was actually used up until now. This condition and the compatibility mode should be removed now.

tinwelint commented on enterprise/metrics/src/main/java/org/neo4j/metrics/source/cluster/NetworkMetrics.java:
```diff
@@ -64,7 +64,7 @@ public void start()
         monitors.addMonitorListener( masterNetworkStoreWrites, ToNetworkStoreWriter.class.getName(),
                 ToNetworkStoreWriter.STORE_COPIER_MONITOR_TAG );
-        monitors.addMonitorListener( slaveNetworkTransactionWrites, MasterClient210.class.getName() );
+        monitors.addMonitorListener( slaveNetworkTransactionWrites, MasterClient214.class.getName() );
```
Shouldn't this be the current version instead?

tinwelint commented:
@chrisvest @burqen had a couple of minor more comments, we're getting there...

chrisvest committed:
Move HA protocol version legend to the newest MasterClient

chrisvest committed:
Remove compatibility mode for HA ID generators, since it was only needed during the 2.0 -> 2.1 upgrade

chrisvest committed:
Point HA network traffic monitoring at the 3.1.0 MasterClient

chrisvest commented:
@tinwelint Comments addressed!

chrisvest commented:
@tinwelint Note that the build actually succeeded, but CI instability caused the PR to be marked as failed.
